### PR TITLE
fix issue 1327

### LINKED ISSUE #1327: [Plato imports] As a librarian, I can follow up on all the to-be-claimed dissertations of researchers
# Enhancement
As a librarian, I can follow up on all the to-be-claimed dissertations of researchers.
Screenshot 2024-04-26 at 11 00 12

Prototype: https://booktower-ugent-ui.vercel.app/biblio/reviewer/reviewer-suggestions.html

### Overview
- [x] Before dissertations are suggested in Biblio, a duplication check happens on title or identifier whenever possible. If the dissertation is not in Biblio yet, they are suggested to reviewers (see next tasks).
- [x] As a reviewer, I can see which records are automatically suggested from other platforms
  - [x] In a separate tab "Suggestions"
  - [x] When there are suggestions, I can see a notification appear in the sidebar (in line with dashboard)
  ```
    <span class="c-sidebar__icon">
      <i class="if if-draft if--notification if--notification-warning"></I>
    </span>
  ```
  - [x] Paginated (in line with publications and datasets)
- [x] I can see the following metadata:
  - Source (Plato image, see image in comments.)
  - Biblio status (Biblio suggestion via Plato)
  - Type (e.g. Dissertation)
  - Access levels (in line with overview pages)
  - Title
  - Publication year
  - Publisher
  - [x] Authors, supervisors, editors, ...

#### 

Example implementation:
```
<ul class="list-group list-group-flush">
    <li class="list-group-item bg-transparent">
        <div class="w-100">
            <div class="c-thumbnail-and-text align-items-start w-100">
                <div class="c-thumbnail c-thumbnail-1-1 c-thumbnail-img c-thumbnail-small c-thumbnail-lg-large">
                    <div class="c-thumbnail-inner">
                        <img src="path-to-images-/plato-logo.svg">
                    </div>
                </div>
                <div class="c-thumbnail-text">
                    <div class="hstack-md-responsive align-items-start gap-3 w-100">
                        <div class="vstack gap-5">
                            <div class="vstack gap-2">
                                <div class="d-inline-flex align-items-center flex-wrap">
                                    <span class="badge rounded-pill badge-default me-3 my-2">
                                        <span class="badge-circle"></span>
                                        <span class="badge-text">Biblio suggestion via Plato</span>
                                    </span>
                                    <span class="c-subline text-nowrap me-2 pe-3 my-2 border-end">Dissertation</span>
                                    <span class="c-subline ps-2 my-2">Embargoed access</span>
                                </div>
                                <h4 class="list-group-item-title mb-0">Using frequency tagging to investigate social processing in autism</h4>
                                <ul class="c-meta-list c-meta-list-inline">
                                    <li class="c-meta-item">2023</li>
                                    <li class="c-meta-item">Ghent University. Department of Experimental clinical and health psychology</li>
                                </ul>
                                <div class="c-author-list">
                                    <div class="c-author">
                                        <i class="if if-ghent-university" data-bs-toggle="tooltip" data-bs-placement="bottom" aria-label="UGent author" data-bs-original-title="UGent author"></i>
                                        <span>Danna Oomen</span>
                                    </div>
                                </div>
                            </div>
                        </div>
                        <div class="c-button-toolbar flex-row-reverse flex-lg-row">
                            <a class="btn btn-link btn-link-muted" href="#" type="button" data-bs-toggle="modal" data-bs-target="#deleteSuggestion">
                                <div class="btn-text">
                                    Hide suggestion
                                </div>
                            </a>
                            <a class="btn btn-outline-primary" href="#">
                                <div class="btn-text">Import as draft</div>
                            </a>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </li>
</ul>
```

#### Preview

https://github.com/ugent-library/biblio-backoffice/issues/1411

#### Importing
- [x] I can import a suggestion as a draft
  - This is a regular draft, as already implemented (new forms out of scope).
  - Any involved authors or supervisors can see the draft / published record in their search overview.
  - Drafts and published dissertations also appear in the "Recent activity" on the dashboard for the person adding the dissertation and the authors / supervisors connected to the dissertation.
image
image

#### Hiding a suggestion
- [x] I can hide a suggestion and get a fail-safe modal
- [x] align/adapt copy
Screenshot 2024-04-26 at 11 00 37

- [x] We can always, behind the scenes, see which suggestions have been flagged as a duplicate, or have been hidden. We can retrieve them if needed.

```
<div class="modal show" id="deleteSuggestion" tabindex="-1" role="dialog" aria-modal="true">
    <div class="modal-dialog modal-dialog-centered" role="document">
        <div class="modal-content">
            <div class="modal-body">
                <div class="c-blank-slate c-blank-slate-muted">
                    <div class="bc-avatar">
                        <i class="if if-alert"></i>
                    </div>
                    <h4>Are you sure you want to hide this suggestion?</h4>
                    <p>You cannot undo this action.</p>
                </div>
            </div>
            <div class="modal-footer">
                <div class="bc-toolbar">
                    <div class="bc-toolbar-left">
                        <div class="bc-toolbar-item">
                            <button class="btn btn-link">No, cancel</button>
                        </div>
                    </div>
                    <div class="bc-toolbar-right">
                        <button class="btn btn-danger">Yes, hide suggestion</button>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
```

### Empty state

- [x] When there are no more suggestions left, there is an empty state:
Screenshot 2023-12-19 at 12 26 42

https://biblio-backend-ugent-prototype-ugent-library.vercel.app/biblio/reviewer/reviewer-suggestions--empty.html

Example implementation:
```
<div class="u-scroll-wrapper__body w-100 p-6">
    <div class="c-blank-slate c-blank-slate-default c-blank-slate-large">
        <img class="mb-4" src="/images/inbox-illustration.svg" alt="" width="auto" height="32">
        <h3 class="c-blank-slate-title">No suggestions, all caught up!</h3>
    </div>
</div>
```

### Optional
- Search (in line with publications and datasets)
- Sort (in line with publications and datasets, keep in mind https://github.com/ugent-library/biblio-backoffice/issues/1364)

## Out of scope
- Flagging of duplicates or faulty suggestions (later)
- Tabs to differentiate between "to import", "imported", "rejected" (by whom and why)

## Context
- This is a first small step in rolling out Plato suggestions
- Administrative support will not be able to see these import suggestions, as we see the researcher as responsible to upload information into Biblio. However, not every research group is organised for this, and administrative support takes over this task. We need our reviewers to be a fail-safe here.
- Groups that are organised differently still are at risk of losing track of dissertations when people have left. We need our reviewers to be a fail-safe here.

### PR DESCRIPTION
**Pull request by nicolasfranck**
fixes #1327 

Done:

* added new page `/suggestions` for reviewers where they can decide to import or delete the suggestion.
* Added new database attributes `status` to mark current lifecycle of the candidate record:
  * `new`: record is available for suggestion
  * `imported`: record has been imported into a publication
  * `deleted`: record has been "deleted"
* Added new icon page `/suggestions-icon` that shows presence of suggestions, and is requested by htmx every 10s, just like `/dashboard-icon` was implemented.
* Added controller to import suggestion. Publication ID is assigned on import, not when it is imported from Plato. Attribute `status` is set to `imported`.
* Added controller to delete suggestion. Attribute `status` is set to `deleted`
* Added controller to confirm deletion before applying it.
* changed `recordsources.Source` interface to make importing "streamable"
* made changes to plato importer:
  * added some attributes to `recordsources.Plato` import as the import failed due to missing required attributes.
  * best to use methods `AddFile` and `AddAbstract` instead of just appending to their array, as these objects require an ID that is assigned automatically.
  * added log entry to trace progress
* made ID and Status attribute in CandidateRecord available. ID is needed for to build controller for specific record
* added the following methods to `repositories.Repo`:
  * `GetCandidateRecords(ctx context.Context, start int, limit int) ([]*models.CandidateRecord, error)` for paged response of "new" candidate records. This name was already long enough ;-)
  * `CountCandidateRecords(ctx context.Context) (int, error)` that counts "new" candidate records
  * `GetCandidateRecord(ctx context.Context, id string)` that returns a specific record, but only if "new"
  * `DeleteCandidateRecord(ctx context.Context, id string) error` that "deletes" a candidate record
  * `ImportCandidateRecordAsPublication(ctx context.Context, rec *models.CandidateRecord, user 
  *  models.Person) (string, error)` that imports embedded data as a new publication with status "private", and a new ID. Note that I use a transaction to move the data from one table to another, and had to fix the `tx` function where the `queries` object was not copied.
* added necessary templ files for all of the controllers. This is getting a bit messy in my opinion to put everything in one folder. But if I want to use shared template only functions (like search_pagination.templ), I have to import that views package (if e.g. I create a `views/suggestions` package). Willing to rewrite this structure

Notes:

* for the cleanup of unused candidate record functionality (like assigned user id, and all of the code that suggest candidate records for researchers) I'll make a separate PR. This PR is already big enough to oversee the intent.

### CODE REVIEW

File: repositories/candidate_records.go [hunk 1]
```diff
@@ -17,28 +18,116 @@ func (r *Repo) AddCandidateRecord(ctx context.Context, rec *models.CandidateReco
...
+
+	err := r.tx(ctx, func(r *Repo) error {
+		pub := rec.AsPublication()
+		pubID = ulid.Make().String()
+		pub.ID = pubID
+		if err := r.SavePublication(pub, user); err != nil {
+			return err
+		}
+
+		if _, err := r.queries.SetStatusCandidateRecord(ctx, db.SetStatusCandidateRecordParams{ID: rec.ID, Status: "imported"}); err != nil {
```
Comment (nics) [inline repositories/candidate_records.go:125]: `SetCandidateRecordStatus` is better

File: routes/register.go [hunk 2]
```diff
@@ -272,8 +273,12 @@ func Register(c Config) {
 				r.Get("/drafts-to-complete", handlers.DraftsToComplete).Name("drafts_to_complete")
 				// dashboard recent activity component
 				r.Get("/recent-activity", handlers.RecentActivity).Name("recent_activity")
-				// dashboard candidate records component
-				r.Get("/candidate-records", handlers.CandidateRecords).Name("candidate_records")
+				// all candidate records
```
Comment (nics) [inline routes/register.go:276]: `r.Use(ctx.RequireUser)` -> this should be restricted to curators only

Comment (nicolasfranck) [reply routes/register.go:276]: you mean to say that `/candidate-records/*` should be restricted or all the existing controllers under `r.Use(ctx.RequireUser)`?

File: repositories/candidate_records.go (hunk 1 above)
Comment (nicolasfranck) [reply repositories/candidate_records.go]: done

File: routes/register.go (hunk 2 above)
Comment (nicolasfranck) [reply routes/register.go]: restricted to user role "curator"

**Comment by mietcls** [comment]
Path is now /candidate-records

**Pull request merged**